src/main/java/net/fabricmc/loom/configuration/providers/mappings/file/FileMappingsSpec.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2022 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.configuration.providers.mappings.file;
+
+import java.util.Objects;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.fabricmc.loom.api.mappings.layered.MappingContext;
+import net.fabricmc.loom.api.mappings.layered.MappingsNamespace;
+import net.fabricmc.loom.api.mappings.layered.spec.FileSpec;
+import net.fabricmc.loom.api.mappings.layered.spec.MappingsSpec;
+import net.fabricmc.mappingio.format.MappingFormat;
+
+public record FileMappingsSpec(
+		FileSpec fileSpec, String mappingPath,
+		String fallbackSourceNamespace, String fallbackTargetNamespace,
+		@Nullable MappingFormat mappingFormat,
+		MappingsNamespace mergeNamespace
+) implements MappingsSpec<FileMappingsLayer> {
+	@Override
+	public FileMappingsLayer createLayer(MappingContext context) {
+		return new FileMappingsLayer(fileSpec.get(context), mappingPath, fallbackSourceNamespace, fallbackTargetNamespace, mappingFormat, mergeNamespace);
+	}
+
+	// Enums don't have a consistent hash code
+	@Override
+	public int hashCode() {
+		String mappingFormat = this.mappingFormat != null ? this.mappingFormat.name() : null;
+		String mergeNamespace = this.mergeNamespace.name();
+		return Objects.hash(fileSpec, mappingPath, fallbackSourceNamespace, fallbackTargetNamespace, mappingFormat, mergeNamespace);
+	}
```

Comment:
Humm, thats a bit annoying, Might be best to use a string name? A string is easier to use from the build script as well.

This applies for both the format and namespace, it also makes a bit more resilient to possible changes in MappingIO.

Reply:
The mapping format is only internal so I'd say it doesn't matter that much. Really, I could just change it to `boolean enigma` since it's either `MappingFormat.ENIGMA` or null.

The merge namespace needs to be exposed so it might actually be nicer to also accept a string and just validate that it's either `official`‚ `intermediary` or `named`, then convert it to a `MappingsNamespace` in `getSourceNamespace`.

Reply:
Yes, that a good idea, a boolean is fine 👍 And I think using a string and validating it would be cleaner.

Reply:
Now it supports both a `MappingsNamespace` (since it's safer and an api enum so people can use that) and a validated string (both are internally a string). The Enigma thing is a boolean now.